src/gadgets/ecc.rs

```diff
@@ -160,6 +160,74 @@ impl<C: CurveAffine<Base = F>, F: PrimeFieldBits, const T: usize> EccChip<C, F,
         Ok(acc)
     }
 
+    // optimization version for scalar_mul
+    // we assume the point p0 is not infinity here
+    pub fn scalar_mul_opt(
+        &self,
+        ctx: &mut RegionCtx<'_, C::Base>,
+        p0: &AssignedPoint<C>,
+        scalar_bits: &[AssignedValue<C::Base>],
+    ) -> Result<AssignedPoint<C>, Error> {
+        let split_len = core::cmp::min(scalar_bits.len(), (C::Base::NUM_BITS - 2) as usize);
+        let (incomplete_bits, complete_bits) = scalar_bits.split_at(split_len);
+
+        // (1) assume p0 is not infinity
+
+        // assume first bit of scalar_bits is 1 for now
+        // so we can use unsafe_add later
+        let mut acc = p0.clone();
+        let mut p = self._double_unsafe(ctx, p0)?;
+
+        // the size of incomplete_bits ensures a + b != 0
+        for bit in incomplete_bits.iter().skip(1) {
+            let tmp = self._add_unsafe(ctx, &acc, &p)?;
+            let x = self
+                .main_gate
+                .conditional_select(ctx, &tmp.x, &acc.x, bit)?;
+            let y = self
+                .main_gate
+                .conditional_select(ctx, &tmp.y, &acc.y, bit)?;
+            acc = AssignedPoint { x, y };
```

Comment:
NBD:
```rust
let sum = self._add_unsafe(ctx, &acc, &p)?;
                                                        
acc = AssignedPoint {
    x: self
        .main_gate
        .conditional_select(ctx, &sum.x, &acc.x, bit)?,
    y: self
        .main_gate
        .conditional_select(ctx, &sum.y, &acc.y, bit)?,
};
```

Reply:
I'll give a little explanation about replacing individual variables with the fields of the initiated struct. This is a good tone, so that the eye does not cling to `let x = `, which without context around it is not clear what it is attached to.